Draping Imagery on 3D Tiles

javagl opened the pull request:
# Description

Addresses https://github.com/CesiumGS/cesium/issues/7591 

The goal of this PR is to allow draping imagery over arbitrary 3D Tiles. 

Currently, tiled image data (like Bing Maps and others) can be draped over the _globe_ in CesiumJS. This is accomplished by defining an `ImageryLayerCollection` for the viewer, causing the "globe (geometry) tiles" to then be textured with the image data. The same should be possible for _arbitrary_ 3D Tiles: Given a (textured or untextured!) 3D Tiles data set, it should be possible to assign imagery layers to that, and drape that imagery over the 3D Tiles geometry. 

### Description of the current state

The following is the attempt of a summary of _the current state, at the time of writing this_. There are several parts that are incomplete or preliminary. Updates will be tracked in comments to this PR.

- There now is a [`Cesium3DTileset.imageryLayers`](https://github.com/CesiumGS/cesium/blob/2bd118a19d50e651c82e7d7693252bf2ec6e0a66/packages/engine/Source/Scene/Cesium3DTileset.js#L1100) that can be used to assign an `ImageryLayerCollection` to a tileset
- The `Model` class has a [_private_ getter for `imageryLayers`](https://github.com/CesiumGS/cesium/blob/2bd118a19d50e651c82e7d7693252bf2ec6e0a66/packages/engine/Source/Scene/Model/Model.js#L1391) that returns the imagery layers of the tileset, **iff** the model is part of a `Model3DTileContent` that is part of a tileset that has imagery layers
 - The model now contains an instance of the new [`ModelImagery`](https://github.com/CesiumGS/cesium/blob/2bd118a19d50e651c82e7d7693252bf2ec6e0a66/packages/engine/Source/Scene/Model/ModelImagery.js) class, which in turn, contains one [`ModelPrimitiveImagery`](https://github.com/CesiumGS/cesium/blob/2bd118a19d50e651c82e7d7693252bf2ec6e0a66/packages/engine/Source/Scene/Model/ModelPrimitiveImagery.js) for each primitive of the model
- These classes both offer `update` functions and `ready` getters, following the pattern that is used for other parts of the model: 
  - The `update` functions are [called in `Model.update`](https://github.com/CesiumGS/cesium/blob/2bd118a19d50e651c82e7d7693252bf2ec6e0a66/packages/engine/Source/Scene/Model/Model.js#L1962), checking for changes that may affect the internal state/data structures, and updating them accordingly
  - The `ready` flags (getters) indicate whether all required resources have been loaded and all computations have been performed
- The `ModelImagery` does not do very much: It mainly delegates the work for the `update`/`ready` handling to the respective `ModelPrimitiveImagery` instances
- The `ModelPrimitiveImagery` is doing most of the work:
  - In its `update` function, it is obtaining the vertex positions of the primitive, from the `"POSITION"` attribute data
  - These positions are converted into _cartographic_ positions for the respective ellipsoid, and stored in a [`MappedPositions`](https://github.com/CesiumGS/cesium/blob/2bd118a19d50e651c82e7d7693252bf2ec6e0a66/packages/engine/Source/Scene/Model/MappedPositions.js) object. This contains the mapped positions and the cartographic bounding rectangle, and serves as the basis for most further imagery-related computations.
  - It computes [`ImageryCoverage`](https://github.com/CesiumGS/cesium/blob/2bd118a19d50e651c82e7d7693252bf2ec6e0a66/packages/engine/Source/Scene/Model/ImageryCoverage.js) objects from these cartographic positions. These describe the (x,y,level)-coordinates of the imagery tiles that are covered by the primitive, and that later have to be draped on the geometry data
  - It currently also processes the state machine of the corresponding `Imagery` objects, to make sure that they eventually reach the `ImageryState.READY` state, so that the whole `ModelPrimitiveImagery` can count as `ready`
  - It creates additional texture coordinate attributes for that primitive - namely, the texture coordinates that will be used for draping the imagery textures on that geometry
  - When all these computations have been performed, the `ModelPrimitiveImagery` counts as `ready`, and can be used as the input for the draping process

The draping implementation itself is implemented as a _pipeline stage_. Summarizing a few things (that should probably be documented on a higher level somewhere):  
- When a `Model` is created, it creates a `ModelSceneGraph` internally
- For each 'primitive' of the model, a `ModelRuntimePrimitive` is created
- The `ModelRuntimePrimitive` contains a list of "pipeline stages" 
- When the draw commands are built (skipping a lot here...), these pipeline stages are executed, and fill `PrimitiveRenderResources` with the data from the `primitive`, which essentially means "copying the data to the GPU, so that it can be used for rendering" 

So now there is an [`ImageryPipelineStage`](https://github.com/CesiumGS/cesium/blob/2bd118a19d50e651c82e7d7693252bf2ec6e0a66/packages/engine/Source/Scene/Model/ImageryPipelineStage.js) that is inserted when the model is part of a tileset that contains imagery data.  

- It prepares the shader that will be used for the draping computations: The shader receives the additional texture coordinate attributes, and a bunch of uniforms that define things like the actual textures as `sampler2D`, and things like the `alpha` value that should be used. All this is then combined in a `sampleAndBlend` function in the fragment shader that takes the previous pixel color, and mixes it with the pixel color that was obtained from the imagery texture
- The actual input for these uniforms is (not yet a proper class, but just) a structure that is referred to as [`imageryInput`](https://github.com/CesiumGS/cesium/blob/2bd118a19d50e651c82e7d7693252bf2ec6e0a66/packages/engine/Source/Scene/Model/ImageryPipelineStageProcessing.js#L173) for now: It contains the `Texture` itself, a rectangle that indicates which _part_ of the imagery texture has to be mapped to that primitive, as well as translation/scale values that are applied to the imagery data so that this rectangle is filled with the right part of the imagery texture

## Issue number and link

https://github.com/CesiumGS/cesium/issues/7591 

## Testing plan

The specs currently only cover the basic aspects of the most important classes and functions. Ideally, _testing software_ should not mean to "check whether it works", but rather to "try and break it". There certainly are many corner cases where one or the other aspect of the current implementation may not work. But due to time constraints, the tests may not be as thorough as one might wish.

Some sandcastles for further tests have been posted in https://github.com/CesiumGS/cesium/pull/12567#issuecomment-2901842522 

The following is a test that was created initially, when this PR was opened,

[testLevels2.zip](https://github.com/user-attachments/files/19713561/testLevels2.zip)

It is a simple tileset with a unit square, with two levels of detail. The first is a plane with 10x10 vertices. The second one are 4 planes, with 10x10 vertices each.

And this is a Sandcastle for testing this on this branch, on localhost (updated for the "Ready for Review" state): 

http://localhost:8080/Apps/Sandcastle/#c=vVhbU9s6EP4rGs5DzUxwwq2XNDCHSznNTCkdAm3p5EW2lUQHW/JIMpCe4b+f1c2XYKDQhocMlrTaXX37aXdFzJlU6IqSayLQDmLkGh0QSYss/GrmgvFKbMYHnClMGRHjldX3YzZm3S46EAQrgtQMfjQlkihEmRlahWMWG/V+ccfNhzImjIS5oBlV9IrIECdJMGYI4WtMlffA/tk8PLPbw4ng2blIwaWZUnm/2015jNMZl6r/ttfb7Doz4b+Ss/FKB/2nNSKUkKiYjmb8ep8XLKFs+pWnRUb6SImCdKxMhm9oVmSjWBDCRjmOyQchuOij9ddG4nZ1zMpjn+SKcoZTOKzMSay4QBP4GTtT0F8BosXdkcmNIiwJ3NHsZOOAQ6/smN5QY6zbtfCVZnRIjnlC0grJuFWF0W8kvc9HdQf7SMNh3KQZnhIxR5HGBgtKJOI2hNOUR8RHsMRwaOW1uf1qy47BEmzRCQoell21YXHuO/Of8JwIwwK354vgVzTR9DPhqdFSazvgXEAggXtyUd7FHGn84aSjeEZ0qGsKvpHomIgYA0xnNZlg1VFBhxohjf/tXZoHhYGYTihJVkvwUuM+CGClN8gZL9IERQThPE9BEilevyRjhuWcxWhSsFgzCcXGgKP5sA5JYODSDoEfh2SCi1ShYF+T7Bjn0rjqaf+LWDaumLX8jYvUS+9p1yoYpZqngJ8TH3JWlx3pxXDvw+lw79Md9JzTTydeiRJCC+Q7FDgnyX0UnOBUag4i1GDhA5tWfYp4IoTPIuRvUNLDaoGFUcXNPSnplDUQ9Wxc5JyJ+0NU87lihFkSY6lSgs6HCOSK3OeBq1oGMifDbJqSQwKR7RlnJWRk8h3yZm14UQ5xms8wjMwgEnQ6U4xI6WfAhhJg2I9nBSn1YlUIrG+LX5ziLHO6br3n53nib2qmvTzGStAbxCd1KDoowhIuJbd79Io/Rde6b/9c+H360FbfmJV3tjCmHJDHlbHAkcri5RUDXJ+LLAI+lAiGfs3GtCZ/ihOQd9wArbNQcZijmMmgbZN1us2EXbkje3Gv7IW/uVZaA3UmwC7Ut6zyqZySIYFwfeZCzc7zM35Eb0hyJDBQ19LVF3Is4GaA/5umhoP7UGZlsPZmO1zf3tjqve2gzXfhu63X795sd9B6I4XU3HbRrEMD4y2jc6QF7laLynJgseg4BMBM2PP3q7bB6QxafBBcGQJakc1FP+zZTp3QD59tXUDvs7T5uKX2A3tDzs6Cc086WP2q3LEVp5yRYGFyePjh89nw7MIya2ExgxpF83Qe1PR2mlTq1G0+QUnzlM/VUiPTHRX1gtrEpTaq9Qa1jANblC7MXEKeS9uaA4QFQYmpR2Zzs+BVeemxrOPLrzUofcrRZa+9kqWETdUM7ezsoF5Z9AQkdsF8QfFkMO7qhq5V0ZSooOdxMqKhyemt+U0v2LBYySrht4lXq/U9viS07fBrdXkoGW2iMF2XqqpJayYsV+t7TL1pEzcLjX4Rwg9xvJ6RxUiieAYJQb8LFiJqNB3BtdYtvo/n/SXmfbV8hws1R0ZFJGNBoRF1zUCdU+6qXDIeX/JChQBlfFkdS+txrzfO0whrTiQ8hocTU5oGH1KiP/fnQ3i5rTgZ+zZc1Ky74Dn0rIlxsLTQ8Zr1Hv2ACqw9BvVDvyMrVyp61yKK5ck1gzYrJ0LNA71ptaT2gge+i9J+n0SSiCscQbWoeWK2l2LS4xa0RKfZgj0YIxOEH5xnC81YB11TuI4YyUxnCj6Z6Nex5OULAkGnRiU8ENJ0DjhICu76cDjpxlv9I8Ea3C9UxbNTzbCgmQzrHcTaxka4bSvfowIbvV7Ysy9f92D7Cac540F5EOuNFljprAzMa2FX7/ybZjm0BaiA93oYdhXJ8lS3x92oiC8hrcT+mv/l6eVCFwELp0K36H0kphEOtjY6yP964Vu7C6EcunPzotjKb9xUBE04EWsCzlHIauG2YYayvFDe2BVwh0I1WMMpdNF9lNEkSUnTwprieR9tlFb8dMSV4lm1AmYGXQ/AIKFXiCY7Lf89QXGKpYQVHd0R/UnGK7uDLsg3tqXcRPQEPIT0o0Vm67uf7GQYhoMuDO/uKq+hicFAaZrvWrcHKuLJfNdzfKDEbvUyGahkd891loMuDJpL1QjGFkA1zwkYFJpscKaMMhj19Be+ga/N1/pbKpLDYB0+4ZbgtQhyAIyvcKobe9/KdpCZsJW0j14ZA6/cGdrNKnKjtH5AD0bbD+tvaGqcDgbiAURMT/n9N/BYL/FY7z2Oh29Ql4OG1f57WFy8LBYXS8Xi4tlY7OnG5o9ck/UKCEi0G/fwWJtb2iXRyp+NxH7Zs/2ZrPErcFR94rIwqSw8G5gD15q+HCy+GV4WKF7/syH5WJCXQwP6/WUBAaqfn0XLl8XLQVG9ZpaWTksLzwbmH/2GejlMzJNtWXAY5b+EBHyWTRl8u27Nt3X/Aw

It offers some basic functions for manipulating the geometry and the imagery settings, targeted at basic tests.

### Notes:

The commit history may look odd: At some point, I started doing the development on a local branch, with fine(r)-grained commits, but for the first time in my life, used the `--squash` function. There just had been too many unknowns and experimental, intermediate steps, and it just wouldn't make sense to track them in all detail here...

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

javagl commented:
The last commits

- The transform of the primitives _within_ the model had not been taken into account. So now, it is passing the `ModelRuntimeNode` to the `ModelPrimitiveImagery` as well, so that it can integrate this transform for the computation of the primitive positions that are supposed to be mapped to the ellipsoid
- The function that obtains the primitive positions did not handle quantized attributes. Basic support for quantized positions is implemented now (and I tested this with a draco-compressed/quantized tileset), but it might have to be reviewed in terms of possible ~"other forms of quantization"

There seems to be an issue related to the computation of the offset/scale for the imagery that is mapped to the primitives. It does work for the test tileset, but not for another ("real-world") one. This might _somehow_ be connected to that quantization, but is hopefully only a detail 🤞 - to be investigated.

ggetz commented:
Thanks for opening this @javagl! I'm looking through the code, but I also has a few higher-level thoughts as I do so:

> The decision about how to pick the right level (of detail) for the imagery is currently still open. There's that XXX_DRAPING_LEVEL slider in the test, which is a hack for testing different levels.

If I understand correctly, the is driven by a maximumScreenSpaceError tolerance in the existing terrain and imagery engine. Can we do something similar here? Something along the lines of refining to the level such that:

`minTileDimensionInMeters / minTileImageDimensionInPixels  < tileset.maximumScreenSpaceError` 

? 

> The function that obtains the vertex positions from the primitive currently copies the attribute data to the CPU. And this does not properly handle quantized data. This might be easy to generalize (if things like Draco or Meshopt do not have to be handled manually). But it might involve shuffling some larger code blocks from existing classes around.

Certainly open to a small refactor to make this easier. If needed, maybe consider a separate PR to keep this scope of this constrained as much as possible.

> How to handle the case that the draping involves more than 16 texture units. This is largely open. It may involve that ominous "upsampling", or maybe could be handled by combining several imagery parts the reduce the necessary number of texture units...?

Am I correct in my understanding that this can be addressed either by:

1. Implementing upsampling: Loader higher LODs of geometry such that we don't have multiple tiles of one imagery layer covering a single geometry tile.

or

2. Merging multiple textures for a geometry tile into one single texture

?

javagl commented:
> If I understand correctly, the is driven by a maximumScreenSpaceError tolerance in the existing terrain and imagery engine. Can we do something similar here? Something along the lines of refining to the level such that: ...

There are two different possible approaches. In CesiumJS, that level is computed [here](https://github.com/CesiumGS/cesium/blob/cf47252a333e6bcf6b2eef51c20daf23f9f24bdd/packages/engine/Source/Scene/ImageryLayer.js#L779) (and I did not yet go through the math and/or justifications there). In `cesium-native`, it is computed with [`computeDesiredScreenPixels`](https://github.com/CesiumGS/cesium-native/blob/5b68acc3c6be0b7f2fb571af420525cdb33eaa32/CesiumRasterOverlays/include/CesiumRasterOverlays/RasterOverlayUtilities.h#L170) ([implementation](https://github.com/CesiumGS/cesium-native/blob/5b68acc3c6be0b7f2fb571af420525cdb33eaa32/CesiumRasterOverlays/src/RasterOverlayUtilities.cpp#L605)), which carries a few implementation notes (and is invoked in some non-trivial contexts, including the translation of these "screen pixels" to the "level"). 

How all this could be translated to the draping is not yet clear. But when you mention the **tileset** `maximumScreenSpaceError`: Note that in `cesium-native`, the decision is based on the [`RasterOverlayOptions.maximumScreenSpaceError`](https://cesium.com/learn/cesium-native/ref-doc/structCesiumRasterOverlays_1_1RasterOverlayOptions.html#a617e8dd5e92f58a998c5dbcf95c5b0b2), which apparently can be configured independently. Whether or not this can simply be "replaced" with the tileset  `maximumScreenSpaceError` in CesiumJS 
 (and maybe throwing in some `/ 8.0` at the right place) has to be examined...


> > The function that obtains the vertex positions [...] does not properly handle quantized data. 
>
> Certainly open to a small refactor to make this easier. 

This turned out to not be such a huge code block and has already been implemented by now. The `AttributeCompression.dequantize` is 'static' and could just be called. What's left is the offset/scale handling. It would be nice to have that encapsulated similarly, because ... my initial implementation was wrong, and fixed in a subsequent commit: There are quite a few `quantizedVolume....`-variables floating around, and _reading_ what they _are_ takes orders of magnitude more time than just calling some nice, opaque, convenient `nowReallyDequantizeThatStuff`-function...

> Am I correct in my understanding that this can be addressed either by:
> - Implementing upsampling: Loader higher LODs of geometry such that we don't have multiple tiles of one imagery layer covering a single geometry tile.
> - Merging multiple textures for a geometry tile into one single texture

These options had been my initial thought. But based on the recent discussion and comments by Kevin, the second one would not be viable. So apparently, there _has_ to be some sort of upsampling. Unfortunately, this is not only about "loading higher LODs", but ... about _creating_ higher LODs (i.e. slicing-and-dicing the actual _geometry_). This raises _many_ questions, but I'll pragmatically start with trying to pick some code fragments from `cesium-native` (which have been ported from CesiumJS, and port them back to CesiumJS ... 😬) and see whether I can use that, for example, to split the unit square into 4 parts.

ggetz commented:
> How all this could be translated to the draping is not yet clear. But when you mention the tileset maximumScreenSpaceError: Note that in cesium-native, ... [this] can be configured independently. Whether or not this can simply be "replaced" with the tileset maximumScreenSpaceError in CesiumJS has to be examined...

Gotcha. From native, it looks like the tileset's geometric error and maximum SEE are factors here, but as you said, ultimately there is a separate SSE option used to drive the imagery refinement (see comments below). So that boils down to

`desiredScreenPixels = projectedRectangleSizeInMeters * tilesetMaximumScreenSpaceError / tileGeometricError`

and we are then looking for the imagery level such that 

`desiredScreenPixels <= imageryMaximumScreenSpaceError`

I am open to having a separate SSE factor for the raster imagery layer— or a modifier that can be applied to the tileset's `maximumScreenSpaceError`— but if there's already a precedent with the former, that might be best for consistency's sake.

---

From native:

```
   * In 3D Tiles, we can only get so close to a model before it switches to the
   * next higher level-of-detail by showing its children instead. The switch
   * distance is controlled by the `geometric error` of the tile, as well as by
   * the `maximum screen space error` of the tileset. So this method requires
   * both of those parameters.
```

```
  // It works like this:
  // * Estimate the size of the projected rectangle in world coordinates.
  // * Compute the distance at which tile will switch to its children, based on
  // its geometric error and the tileset SSE.
  // * Compute the on-screen size of the projected rectangle at that distance.
  //
  // For the two compute steps, we use the usual perspective projection SSE
  // equation:
  // screenSize = (realSize * viewportHeight) / (distance * 2 * tan(0.5 * fovY))
  //
  // Conveniently a bunch of terms cancel out, so the screen pixel size at the
  // switch distance is not actually dependent on the screen dimensions or
  // field-of-view angle.
  ```

```
   * The `target screen pixels` returned here may be further modified by the
   * raster overlay's {@link RasterOverlayTileProvider::getTile} method. In particular, it
   * will usually be divided by the raster overlay's `maximum screen space
   * error` of the raster overlay (not to be confused with the `maximum screen
   * space error` of the tileset, mentioned above).
```

ggetz reviewed (commented):
A few comments on the code so far...

javagl commented on packages/engine/Source/Scene/Cesium3DTileset.js:
```diff
@@ -2811,7 +2830,8 @@ function addTileDebugLabel(tile, tileset, position) {
 
   if (tileset.debugShowGeometricError) {
-    labelString += `\nGeometric error: ${tile.geometricError}`;
+    // XXX_DRAPING Show SSE as well....
```
I'll leave this open, and during a "cleanup pass" where all the `_DRAPING` comments are handled, this can be moved into an issue (there might be other "minor TODOs" ... and some larger ones as well ... coming out of all this)

javagl replied:
This will be more relevant for upsampling, so I moved this into https://github.com/CesiumGS/cesium/issues/12628 and will resolve this for now.

javagl commented on packages/engine/Source/Scene/Model/ImageryCoverage.js:
```diff
@@ -0,0 +1,70 @@
+   */
+  constructor(x, y, level, textureCoordinateRectangle) {
+    /**
+     * The x-coordinate of the imagery tile
```
I'm usually documenting and commenting _extensively_ (some might say _too_ extensively). But sometimes, I have to rely on documentation that is already _there_. You cannot imagine how much headscratching and confusion was involved in some places, to figure out whether some "north" value is _smaller_ or _larger_ than some "south" value. I'd like to add stuff like that, but ... want to be sure (wrong comments are worse than no comments...)

javagl commented on packages/engine/Source/Scene/Model/ImageryCoverage.js:
```diff
@@ -0,0 +1,70 @@
+     * @type {Cartesian4}
+     * @readonly
+     */
+    this.textureCoordinateRectangle = textureCoordinateRectangle;
```
Yes, I've occasionally been meandering between `Rectangle`, `BoundingRectangle`, `Cartesian4` (and added a local `CartesianRectangle`), because they serve different purposes in different contexts, with different constraints. Here, the main purpose was to ~"have something that can be passed to a shader via a uniform", but that translation may also happen elsewhere.

javagl replied:
In the last state, I replaced that with a `CartesianRectangle`. When this is about to be sent to the shader, it has to be done as a `Cartesian4`, but I think that this makes some sense: The `ImageryCoverage` is a somewhat "generic" representation of _that imagery_, and should use the rectangle here. Later, when this information is about to be sent to the shader, these "coverages" will be converted into an `ImageryInput`, which then _does_ contain the shader-oriented representation (including that rectangle, but there, as a `Cartesian4`).

javagl commented on packages/engine/Source/Scene/Model/ImageryCoverageComputations.js:
```diff
@@ -0,0 +1,431 @@
+// be abstracted away using a `MapProjection` and/or computing only the
+// right texture coordinates for the respective projection to begin with.
+
+// TODO_DRAPING Preliminary, internal class for texture
+// coordinate and index range computations. This could
+// be emulated with a "BoundingRectangle", but min/max
+// is often more convenient than min/size.
```
In the meantime, I actually _extended_ that `CartesianRectangle` class, to offer some baseline functions like `bool containsInclusive(x,y)` ... Maybe all this can be moved to `BoundingRectangle` at some point.

javagl commented on packages/engine/Source/Scene/Model/ImageryCoverageComputations.js:
```diff
@@ -0,0 +1,431 @@
+    // Similarly, if the northwest corner of the rectangle lies very close to the south or east side
+    // of the northwest tile, we don't actually need the northernmost or westernmost tiles.
+    // We define "very close" as being within 1/512 of the width of the tile.
+    const veryCloseX = inputRectangle.width / 512.0;
```
I have no clue. This was taken from https://github.com/CesiumGS/cesium/blob/7f57ecddce231afffec77240b30065b99beb5e95/packages/engine/Source/Scene/ImageryLayer.js#L821 and https://github.com/CesiumGS/cesium/blob/7f57ecddce231afffec77240b30065b99beb5e95/packages/engine/Source/Scene/ImageryLayer.js#L900 . A wild guess is that it is 

1. the assumption that imagery has 256x256 pixels
2. the goal to let this be "half a pixel"

(It might be that all this "veryClose"-handling turns out to be obsolete, or solved more explicitly. But for now, I have to say "it seems to do the right thing (let's not touch it)....")

javagl replied:
Probably not actionable item here. (Except for reviewing that whole "veryClose"-stuff, but that's not specific enough to be actionable right now)

javagl commented on packages/engine/Source/Scene/Model/ImageryPipelineStage.js:
```diff
@@ -0,0 +1,706 @@
+
+    //>>includeStart('debug', pragmas.debug);
+    Check.defined("modelPrimitiveImagery", modelPrimitiveImagery);
+    if (!modelPrimitiveImagery.ready) {
```
This should not be called when the modelPrimitiveImagery is not present. But there are several places where I wondered about 1. what should be checked (e.g. with `Check` or manually), or 2. what could be established as a _contract_ (precondition) for some `@private` function. During development, it doesn't make much of a difference whether something is a `DeveloperError` or just throwing because of some `undefined`-acess....

javagl replied:
This actually became obsolete with the latest changes. As discussed, we'd like to have the option to "continue with everything (and count the model as `ready`)" even when the imagery is _not_ ready yet. So this case had to be handled here accordingly. Fortunately, that "handling" just consists of `if (not ready) return;`.

javagl commented on packages/engine/Source/Scene/Model/Model.js:
```diff
@@ -1375,6 +1378,33 @@ Object.defineProperties(Model.prototype, {
+   */
+  imageryLayers: {
+    get: function () {
+      // TODO_DRAPING: Currently, this implies that a Model only can have
+      // imagery when it is part of a tileset. One could make that case
+      // that it should be possible to assign imagery to a standalone
+      // model. This could probably be implemented here, with reasonable
+      // effort.
```
This comment/idea may be obsolete. From the latest discussion around "upsampling", it looks like the draping can not happen on the `Model`-level alone. (Or at least, not the way that it did until now). I'm afraid that it will be necessary to turn that "`Model` with `Imagery`" into a full-fledged, small "tileset" that starts at this node 🙁 ...

javagl commented on packages/engine/Source/Scene/Model/ModelImageryMapping.js:
```diff
@@ -0,0 +1,483 @@
+        attribute.componentDatatype,
+        count * componentsPerAttribute,
+      );
+      buffer.getBufferData(typedArray);
```
Probably yes. 

I should not have removed that "Some generic reader wuld be nice" comment that I originally added there. [Some generic reader could avoid a lot of headaches](https://github.com/CesiumGS/cesium/pull/11842)...

ggetz replied:
I generally agree about a generic reader. Though this may be something else out of scope for the PR. I don't think we have an issue open. Would you mind creating one?

javagl replied:
As mentioned in the comment below: Given that upsampling will make it necessary to read _all_ attributes (regardless of their representation), I think that such a "generic reader" will not only be "nice", but "(unfortunately?) be _required_". I started that with the `ModelReader` class, and assume that (even though that's not fully integrated and tested) this resolves this comment.

javagl commented:
The last commit (which is, again, squashed from a local state) aims at working towards a state that allows for _some_ basic tests. Some details are summarized below.

---

### About "upsampling"

There has been some discussion internally, about the process of "upsampling" the geometry data. I did start some experiments and created some building blocks. But the upsampling is **not** enabled in the current state. There are still too many building blocks missing.

This is roughly related to some of the inlined review comments:

> javagl: "I should not have removed that "Some generic reader would be nice" comment that I originally added there."
> ggetz: "I generally agree about a generic reader. Though this may be something else out of scope for the PR. I don't think we have an issue open. Would you mind creating one?"
> ggetz: "Ah, I meant things like interleaved buffers, quantized vs un-quantized, etc..."

The point of such a "generic reader" would exactly be to provide _some_ form of abstraction from all these variants. Certain tasks that are required for the upsampling are pretty simple, like "GIve me the position/normal/texCoord of a vertex". But there is no infrastructure for that. It will have to take fetch the data from the GPU, and take into account whether the data is normalized, quantized, oct-encoded, draco-compressed, or stored in separate or interleaved buffers (and for indices, whether they are `TRIANGLES` or `TRIANGLE_STRIP`), and eventually, _all_ (float) attributes of the data (regardless of their representation) will have to _read_, subdivided, interpolated, and assembled into "new geometry data". There are many questions related to that. So while there is a draft for some `ModelAttributeReader` in the current state, this is not fully wired in. I locally started creating test data for that, as in files
```
unitSquare11x11_draco.glb
unitSquare11x11_draco_interleaved.glb
unitSquare11x11_meshopt.glb
unitSquare11x11_meshopt_interleaved.glb
unitSquare11x11_plain.glb
unitSquare11x11_plain_interleaved.glb
unitSquare11x11_unsignedShortTexCoords.glb
unitSquare11x11_unsignedShortTexCoords_interleaved.glb
unitSquare11x2_triangleStrip_plain.glb
...
```
But even if this was already tested and implemented properly, it would only be a tiny part of the upsampling. So for now, upsampling is omitted. 

---

### About the current state:

Instead of "upsampling", the process is currently computing a "matching" imagery level for the geometry at hand, and drapes this over the geometry. So it is essentially trying to always map **one** imagery texture to a certain primitive. This does not work for all possible cases. Specifically, it assumes that _the geometry data itself_ provides a sufficient level-of-detail structure. (So when the tileset contains a single LOD with a 1000kmx1000km square consisting of 4 vertices, this won't work well...). But it allows for some basic tests, e.g. with Google Photorealistic 3D Tiles.

The following is a sandcastle for such basic tests when running the current state on localhost:

http://localhost:8080/Apps/Sandcastle/#c=vVZbb9s2FP4rRPZQGbDluxV7TrDEXdsA7RrU7friF0qiJSIUKZCUE6/If98hJUqy6wVFsOTBhshz5XeukeBKox0l90SiC8TJPVoRRYvM/9veeZuzyJ5XgmtMOZGbsy76seEIJUyEZIG2mCnS3fDHzu8bHll1mjKiiAZ9+B5T7TRGkmBN3guRMHKbCi3gzKjSNBq//VqKeFaz4Gz/TVGefKc6LfnfExGJmMgF0rKozZV++yoinPi5pBnVdEeUj+PYq5wwbBve76OVtY50SpBXcJWTiG4piTuIZjghco8Y3hOpgAFrK6BSUbAYhQThPGfAirSw4pXmDcdqzyO0LXikqeCofF/1kptS60er1OuUkB3Ac8BxiH2btBKMEWvAK9+CEDj3lmxxwTTyrgEm9AnnqmMo1go6acGAUl3cSrGjsQmu5UanwvRdSOb4r8w7vR+OGyGl9wxC77wVvM29NkT/6s8vN1cfu07msVN+mSfAqX7HOyFRTMIiSeAdC7ROxb3F2AUlFAWPsaREQVYcou/gtOJG8K3EOXFemDBcN8IXZZ5a63SLvF8R6jwXzlYkjcaVEDKmHEBVxxItTMEIQLCOUpIBtC0V30n4icgIQ718bfF4nafBlUQXkp/0Hdgeq6q4Uoom/AByVwdHyV6CTY+StsycpzIfjFU6/GNheuSU9WiNeRxhpRlB324QiBW5M27K/RN0AQayFjnM8hQv0NAiEUqapJoTpdwNSGkJqtw5LQDZgf1UGNDBpqocMcFZVul6dL58y2PXM8ARDdhDGm7BLDvVNRCWBMU2n6z0YcJ2UYgVNJHq1rwFZeYxG153kMLac3VUWXTNw6TtAWI+IzzRKbq4uECDOlnLuLtMcEVi3TyG3E+g4w5cV7EsvoUUGP8qshAStIa8JJQZVnI2eJ9ib6htGReRUxKO1uaHiJ1ihes2VxPMU8wNtS1jw32K3RI6rRJZQbghbvcpOY4cilLME6J+iqDV9E6KzAxRF7//CG7L0roIVSQpTJyq+NpJUnWDOy6iO1FoH7CK7hq/W9NXCBZiE+xYREVGuDZx/pMR83m9v4lhplc8mzMjdqzZjLs9zJXYOlhb6DrNRmYLfdsr7XGcQfPgTXW28rUVMqw+33NofDmReu8ZoU6ds0ceuL5m/P4cKiJ3OGSk7YkVr9mUw807AX/dFB3Kt4JybeGNQIvECJenctHIGXk4WixKNh9q2IbT+hwT2Ft41UBazXqFJbR5ivm4mga90Ww+DuaBP50OB7PBJAiqtt2bjObBcDz1p8NgMJ8Eo2lFGJ9PJ8Fg4AfwPxlOxrOJuS+7vYC5BHvYT2Y/EGyidUt1lH6BhaEyPvFn0+kA7M5H0+FwPq9t9Ab+6Hw2Pp9P4H84Pj8fzyrK0B9PZ8MAPB1Ng2A8DMakFzgH7NJ11j1b2gXg0lz/QbNcSI0KyTzf72uS5cxMun5YRHfQ8yPXAH5zeVnFPIT0TaQZtgskkxB7k1EXud/APy+lEMph0NrdYJI/VFchzFMiexLeXKiG8HhghvK80M7YDpKORpj1YOFMALqMxjEjhxZ6WuQLNKqtuOtQaC2yhgJmln0HwDKmO0TjixNrMooYVgoo24KxNf2HbM4ul33gPxBjwkbuM3gIjcmwpMPLj+Wl7/vLPhx/lqrr18ZgqU19XJZuL3Uo4v2lK46llpfNkrHU8eWVaeTLPnwd3jcnOJfo6X1OwJo0XQ4elFEOp4H5wg/wNYQvpUluLv3BCE5QfbgXQuuAqx1mZt7audFF9lTO0wV6Y9W/qdw/bVSTB230A3Bwmj6h/EDNwcPgIJ9A4rqeUf8LHONfgaOZiy+FSWPh2cCsqlH8erC44f9SoDj9z4bkQ0FeDw3Yb14KCFD9bAzW9Sb1elA029tLIdJYeDYw783O+HqY2BX1peCwyn8JCfisRw18VzPIDat/AQ

It is using Google Photorealistic tiles, draping Bing Aerial maps over them, and offering some basic imagery layer manipulation UI for testing. A short screencap:

![Cesium Draping P3DT preview](https://github.com/user-attachments/assets/882e68c1-493f-4cd3-af5a-35a93bc4beb6)

---

Given that there are many commented-out blocks related to ~"upsampling experiments" in the current state, it does not make sense to review this state _on a code level_. But maybe someone wants to try it out.

javagl commented:
The last commit added a few cleanups (removing obsolete comments, omitting (hopefully) obsolete debug logs, and omitting the part that is related to the upsampling experiments).

The upsampling might be added in a PR that is built _on top_ of this one, to make the process of reviewing and testing easier.

javagl commented:
There are still many `TODO_DRAPING` comments in the current state. Some of them are rather "notes" for things that may require additional attention during a review. Others are "trivial" cleanups, with minor degrees of freedom (e.g. some of the ones that are related to the inlined comments from your initial review that are still open). 

The "larger" ones refer to some of the points that had been listed as "To do" in the initial comment, often somehow related to upsampling. I just removed that list from the initial comment, here is one with some updates:

- The decision about how to pick the right level (of detail) for the imagery is currently still open. There are some inlined comments about how this is done in CesiumJS/native for quantized meshes, sometimes depending on the SSE for the tileset or the imagery. "The right" solution still has to be assembled. Some of this may be obsolete without upsampling: Even if we compute some level, we can not necessarily _apply_ that level. So I think that the [current approach](https://github.com/CesiumGS/cesium/blob/4c3a45b590a1b9f194ecd009d15c6e2fe7f140bf/packages/engine/Source/Scene/Model/ModelPrimitiveImagery.js#L427) of trying to ~"map roughly one imagery texture to each primitive" is the safest bet.
- Obtaining the vertex attrobutes from the primitive: For the current state, this "only" needs the positions. But for upsampling, it will be necessary to read _all_ attributes. And I will not try to implement this by copy-and-pasting [`getVertexPosition`](https://github.com/CesiumGS/cesium/blob/7f57ecddce231afffec77240b30065b99beb5e95/packages/engine/Source/Scene/Model/pickModel.js#L366) and tweaking it for normalized, ZXY-encoded normals or so. There _has_ to be some abstraction for that. Ideally, this would be an `AccessorView` (as in `cesium-native`), preferably something that also handles [further corner cases](https://github.com/javagl/JglTF/blob/84ce6d019fec3b75b6af1649bbe834005b2c620f/jgltf-model/src/main/java/de/javagl/jgltf/model/AbstractAccessorData.java#L149). For now, I started a [`ModelReader`](https://github.com/CesiumGS/cesium/blob/4c3a45b590a1b9f194ecd009d15c6e2fe7f140bf/packages/engine/Source/Scene/Model/ModelReader.js#L16) (and actually have some specs for that locally), but some work has to go into that.
- The processing of the "state machine" of the `Imagery` has to be reviewed. Part of that is that there's that `Imagery.addReference` function, which is called at a [somewhat random place](https://github.com/CesiumGS/cesium/blob/4c3a45b590a1b9f194ecd009d15c6e2fe7f140bf/packages/engine/Source/Scene/Model/ModelPrimitiveImagery.js#L399) right now. In a similar vein, some aspects of memory management in general have to be reviewed (and maybe I'd add _some_ comment to things like `buffer.vertexArrayDestroyable` once I figure out what this is about...).

If desired, I can try to remove "as many TODOs as possible" (i.e. the ones that are rather "notes" or indicate "small fixes").

ggetz commented:
@javagl It looks like there is an error in `built-ts` which is preventing this branch from being deployed in CI. Having this branch deployed and available for testing would be helpful. Would you mind taking a look at the error?

javagl commented on packages/engine/Source/Scene/Model/ImageryPipelineStage.js:
```diff
@@ -0,0 +1,706 @@
+      // TODO_DRAPING These are short-circuiting to `number` instead
+      // of `boolean` here. Ignore that or do that clumsy `x = x && y`
+      // or some `!!x` trickery...?
+      imageryFlags.alpha |= imageryLayer.alpha !== 1.0;
```
For now, I added a short remark in the `ImageryFlags` class, saying that these properties are `boolean` **or** `number` (and should be checked to be 'truthy' or 'falsy' by users), and opened https://github.com/CesiumGS/cesium/issues/12623 to track possible ECMA updates.

javagl commented on packages/engine/Source/Scene/Cesium3DTileset.js:
```diff
@@ -1097,6 +1097,25 @@ Object.defineProperties(Cesium3DTileset.prototype, {
+    // `tileset._imageryLayersDirty` flag that is passed to the
+    // models, so that the models can be updated for new imagery
+    get: function () {
+      return this._imageryLayers;
```
One question that still has to be answered for that one is how to track changes. The inlined comment causally mentions a "`tileset._imageryLayersDirty` flag", and that is what it will boil down to. But given that the imagery layers in the model are taken from the tileset, **if** the model is part of tile content, raises the question about where to actually track this flag. 

One approach of which I think that it could make sense would roughly be:

- Store some `imageryLayersModificationCounter = 0` in the tileset
- Attach a listener to the `ImageryLayerCollection` that just increases this counter for each modification
- In the model, store some `lastImageryLayersModificationCounterWhereUpdateHappened`
- In the `model.update` function, pragmatically check

```
if (this.lastImageryLayersModificationCounterWhereUpdateHappened != 
    owningTileset.imageryLayersModificationCounter) {
  updateWhateverHasToBeUpdated();
  this.lastImageryLayersModificationCounterWhereUpdateHappened = 
    owningTileset.imageryLayersModificationCounter;
}
```
The reason for asking: I have not seen such a "modification counter"-based approach in other parts yet. Similar things are often handled by copying values to the respective object (model) in each update, and then detecting changes in the values _locally_...

javagl replied:
The "modification-counter" based approach was now added in https://github.com/CesiumGS/cesium/pull/12567/commits/b6d0156afb3997700ffbb60b386039b1892a25a6 , together with some related aspects of lifecycle management. Some questions are still open (e.g. where exactly to increase/decrease the reference counter of `Imagery` objects), but I now implemented something of which I _hope_ that it makes sense. The basic tests consisted of a local sandcastle where I could add/remove an `ImageryLayer` in the `tileset.imageryLayers`. Removing it was tested with `destroy=true` and `destroy=false`. All this still has to be covered with specs, but on the level of the "Integration Test (Sandcastle)", it seemed to work, so it's a first shot, at least.

javagl replied:
(Note that the demo sandcastles in the comments below usually say `tileset.imageryLayers = theNewOnes`, which now is **no** longer possible, but I'll update these sandcastles as part of the cleanup anyhow)

javagl commented on packages/engine/Source/Scene/Model/ModelImageryMapping.js:
```diff
@@ -0,0 +1,483 @@
+    // dequantized data
+    const p = new Cartesian3();
+
+    // XXX_DRAPING: There's also some quantizedVolumeScale floating around in
```
Similar to the comments above and below: Yes, we will need something that can real "all the things", and I started with that (so assume that this comment is resolved by that ongoing effort)

javagl commented on packages/engine/Source/Scene/Cesium3DTileset.js:
```diff
@@ -1097,6 +1097,25 @@ Object.defineProperties(Cesium3DTileset.prototype, {
 
+  /**
+   * The {@link ImageryLayerCollection} that should be draped on the contents of this tileset.
+   *
```
Known limitations are one thing. I'm more concerned about the _unknown_ ones, because these will show up as issues 😬 . But I'm not sure what this _particular_ comment referred to: Eventually, most tile formats will boil down to `Model` (but we can check which ones don't). For example, I already tried out applying the imagery draping to a (PNTS) point cloud, and that worked...

javagl commented:
Some of the last commits had been twiddling with CI builds. We currently are between a rock (TSDoc) and a hard place (ESLint): When a class is only used in the _documentation_, then there are two options: When importing that class, ESLint complains, because for ESLint, it counds as "unused". When _not_ importing it, then TSDoc complains, because that type is not found. We should probably address https://github.com/CesiumGS/cesium/issues/10455 to straighten that out.

---

The main last commit (squashed again) does some cleanups (mainly removing debug outputs etc), and some smaller functional changes:

- The texture coordinate attribute for the imagery was previously computed in the `ImageryPipelineStage`. But given that this may be required for the upsampling, it is now computed in the `ModelPrimitiveImagery`. (The pipeline stage is only responsible for handing that to the GPU via the render resources)
- Minor cleanups for error handling: When an imagery went into the `FAILED` or `INVALID` state, this caused some helpless `"What do do here?"`-messages. Most of these messages are now omitted, because the answer is: Nothing. Failed imageries already create a log message from the ImageryProvider. That should be enough
- The model now does no longer wait for the imagery to be "ready" before itself counts as `ready`. The reason is that during some tests with a slow imagery server, it took ~1 minute until the first tiles appeared, because it had been waiting for the imagery. I think that displaying the model without the draped imagery may still not be ideal: People could think "Hey, it doesn't work!?!", not knowing that they'll just have to wait a minute. So this is currently done with a compile-time flag, but may have to be configurable in the future, and this "configurability" _might_ include some sort of _visual hint_ that it is still waiting for the imagery to become available.

---

Beyond that, some of the internal work aimed at upsampling, or rather at _not upsampling_: I have a few upsampling experiments locally, and one goal was to carve them out in a way that they could be _omitted_ in _this_ PR, and (hopefully) be "hooked in" at the right place in a follow-up PR.

javagl commented:
The TypeScript-generation step has been reactivated. The new classes are marked as `@private`, and that makes the TypeScript generator happy here. The only public, user-facing change of this PR is, indeed the getter for `tileset.imageryLayers`, which is now tagged as `@experimental`.

ggetz commented:
Thanks for the examples @javagl! This is working super well! I'm in the middle of a review pass on the code itself.

A couple comments on the examples:
 - I love the use case where we drape labels on top of Google Photorealistic Tiles– super useful. A note that it likely violates Google's terms of service to use other map data (Bing) alongside of theirs. So, for any published Sandcastle examples, I'd like to check if there is a similar Google raster layer we could possibly use to showcase instead.
 - I noticed due to the "blend" in the shader, when displaying a tileset on terrain with the same imagery, the tileset shows up a slightly different color. Do you know if this is the same blending that gets applied by default in Cesium for Unreal? I think we should try to aim for consistency where it makes sense. 
![image](https://github.com/user-attachments/assets/8dce75b1-d8e3-4727-9cc9-27c60c5242df)

javagl commented:
There are still a few tasks and TODOs, with varying degrees of importance and how much they are related to the "core" of the current implementation. From the tip of my head:

- Specs
- Specs
- Specs
- There's that antimeridian issue (distorted/wrong texture coordinates). It's unclear how much time it will take to investigate that, and to figure out what _eactly_ the fix is. (It might very well be 10 hours of digging and debugging, to eventually find the place where a single `x = clamp(x, 0.0, 1.0)` has to be inserted...)
- Related to the "Bing Labels + Google" point:
  - Organizational: What demos can or should be shown
  - Technical: The credit handling. I haven't looked at the details yet. It _might_ be possible to just let the `ModelImagery` sneak the  `imagery.credits` into the `model.credits`, but maybe that should be handled differently
- The brightness/appearance difference
- Just for completeness (out of scope here)
  - Upsampling
  - Outline rendering
  - ...

About the brightness/appearance: 

I saw that. And there are _many_ differences between how _models_ are rendered and how _terrain tiles_ are rendered. The "core" of the draping rendering, the [`sampleAndBlend`](https://github.com/CesiumGS/cesium/blob/2d972b4a492d4f40ba93c3a77f8a51e3a528db95/packages/engine/Source/Scene/Model/ImageryPipelineStage.js#L467) function, is re-assembled (using a `ShaderBuilder`) from the [original `sampleAndBlend`](https://github.com/CesiumGS/cesium/blob/2640bd8b471a9c6d398d04fd5fa21177411e24ef/packages/engine/Source/Shaders/GlobeFS.glsl#L166) function. But there is **a lot** of random stuff happening in `GlobeFS.glsl` that will affect the _final_ appearance. It's hard to pinpoint "the" exact reason for the difference. I did originally start with [writing comments into GlopeFLS.glsl](https://github.com/CesiumGS/cesium/blob/2d972b4a492d4f40ba93c3a77f8a51e3a528db95/packages/engine/Source/Shaders/GlobeFS.glsl#L90), but ... eventually had to shrug it off and focus on "getting stuff done". What I did notice, however, that this seems to **only** be caused by the gamma. For example, with a gamma of ~0.45, the AGI building is basically indistinguishable from the terrain imagery:

![Cesium Draping Gamma](https://github.com/user-attachments/assets/90602410-ceff-4e46-a802-1a33d9095947)

ggetz commented:
Nice catch @javagl! Are we just missing a [gamma correction](https://en.wikipedia.org/wiki/Gamma_correction#:~:text=In%20most%20computer%20display%20systems%2C%20images%20are%20encoded%20with%20a%20gamma%20of%20about%200.45%20and%20decoded%20with%20the%20reciprocal%20gamma%20of%202.2.) then?

I think so– Calling `czm_srgbToLinear` right after reading the texture in `ImageryPipelineStage.js` seems to do the trick of matching what's happening for the globe:

| Before | After |
| --- | --- |
| ![image](https://github.com/user-attachments/assets/a4645a8b-4ff0-4907-8a6b-4da2001e10f0) | ![image](https://github.com/user-attachments/assets/e860d61d-c9e0-406b-a049-e07654d14419) |

javagl commented:
On the one hand, I'm tempted to spend much more time with verifying things here (~"Are the color spaces actually specified for the imagery?", "Where was this (or something equivalent) done in the Globe shader?", and "Why didn't that ~0.45 catch my attention in the way that it should have 😆 ?"). But on the other hand... it looks right, and time is short, so I added this.

javagl commented on gulpfile.js:
```diff
@@ -626,12 +626,12 @@ export const makeZip = gulp.series(release, async function createZipFile() {
           "Apps/Sandcastle/.jshintrc",
           "packages/engine/index.js",
-          "packages/engine/index.d.ts",
+          //"packages/engine/index.d.ts",
```
Indeed, that was a leftover from disabling the TS build step

javagl commented on packages/engine/Source/Scene/Cesium3DTileset.js:
```diff
@@ -1020,6 +1050,29 @@ function Cesium3DTileset(options) {
+  const imageryLayersListener = function () {
+    tileset._imageryLayersModificationCounter++;
+  };
+  imageryLayers.layerAdded.addEventListener(imageryLayersListener);
+  imageryLayers.layerRemoved.addEventListener(imageryLayersListener);
+  imageryLayers.layerMoved.addEventListener(imageryLayersListener);
+  imageryLayers.layerShownOrHidden.addEventListener(imageryLayersListener);
```
It's not immediately obvous why the `GlobeSurfaceTileProvider` stores one callback for each event, instead of storing the _listener itself_ and then removing it. I took the latter approach here now (Specs should show whether this works as expected...)

javagl commented on packages/engine/Source/Scene/Cesium3DTileset.js:
```diff
@@ -1097,6 +1150,56 @@ Object.defineProperties(Cesium3DTileset.prototype, {
   },
 
+  /**
+   * The {@link ImageryLayerCollection} that should be draped on the contents of this tileset.
```
> Technically, although uncommon, 3D Tiles content can include GeoJSON data and experimental vector data– I believe neither of those will work with overlays.

GeoJSON should work. (I just tried it with basic examples, and... these are just `Model` instances, eventually). Not sure about vector data. I3DM is tricky.... it ""works"", but all instances have the same texture, so it's not really "draped" (based on the instance position).

I now listed the types explicitly:

"The imagery will be draped over glTF, B3DM, PNTS, or GeoJSON tile content."

>Do we also want to document the limitation on level of detail due to upsampling, which will land later?

I'm not sure whether this is the right place. It's may not be "relevant" for users insofar that they cannot change the behavior. One could say something like 
"The imagery level will be picked based on the bounding box size of the cartographic positions of the model primitives"
or so, but ... I think that those who _understand_ this information don't really _need_ it 😆 

> Can we add an @example here? And add a 3D Tiles @example to the documentation for ImageryLayer as well?

I have added an `@example` here. But the `ImageryLayer` has nothing to do with the `Cesium3DTileset`. (Yes, the tileset can _use_ the imagery layers, but ... I'm thinking about the "direction of dependency" here...). 

One could consider to change the existing examples in `ImageryLayer` from
`scene.imageryLayers.add(imageryLayer);`
to
`imageryLayers.add(imageryLayer);`
or even
`imageryLayerCollection.add(imageryLayer);`

Otherwise, that change would be to just _duplicate_ one of the existing examples, and replacing `scene.` with `tileset.`....

javagl commented on packages/engine/Source/Scene/BoundingVolumeDerivation.js:
```diff
@@ -0,0 +1,320 @@
+ * They should preferably be offered for objects, and not only for
+ * their JSON representation
+ */
+class BoundingVolumeDerivation {
```
This class contains functions that have been pulled out from [`Implicit3DTileContent`](https://github.com/CesiumGS/cesium/blob/2640bd8b471a9c6d398d04fd5fa21177411e24ef/packages/engine/Source/Scene/Implicit3DTileContent.js#L789), as one (early) stage of what has to be done for upsampling - i.e. "Split some bounding volume into 4 parts". I'm not even sure whether this will be useful. The functions operate on JSON data. We _might_ need them for the actual _objects_ (but that's not clear yet). 

In terms of the style ... well, maybe that's my Java background. 

Having individual, free-floating functions ("alternative 1" from the link) seems counterproductive for me, in terms of `export/import` handling. 
(These free-floating functions - including the (non-static) 'instance' functions) - will cause lots of headaches for any form of codebase modernization. In _proper languages_, there could be the additional differentiation to make some of these functions acutally `private` (in the strict, language-level sense)) 

And... I don't see a compelling reason to _avoid_ these classes, **iff** the alternative is to just change them into an object (i.e. the "alternative 2" from the link). For the user, it does not make a difference. I could imagine this to actually be translated to the same JavaScript under the hood.

**tl;dr**

1. I can omit this change (because it is not strictly required for this PR)
2. I can change that `class` into an object (although I'm not sure what the benefit is)

javagl commented on packages/engine/Source/Scene/BoundingVolumeDerivation.js:
```diff
@@ -0,0 +1,320 @@
+ * Functions to derive child bounding volmes in quadtrees and octrees,
+ * from the root bounding volume and implicit coordinates.
+ *
+ * These functions have been extracted from Implicit3DTileContent.
+ * They should preferably be offered for objects, and not only for
+ * their JSON representation
```
See comment above: This was part of the "upsampling experiments/preparation". It's not clear whether this is useful. It may be omitted here.

ggetz replied:
Got it. If it's not too much trouble to pull out of this PR, I think that would help with keeping the scope small.

javagl replied:
Done via https://github.com/CesiumGS/cesium/pull/12567/commits/59e2b0357c14a708856e5afddd4e44797ea82bd1

javagl commented on packages/engine/Source/Shaders/GlobeFS.glsl:
```diff
@@ -87,25 +87,48 @@ flat in int v_regionIndex;
+// - The applyFog depends on enableFog, and some tile distance from the viewer
+// - The showGroundAtmosphere is a flag that is passed through from the Globe,
+//   and is true by default when the ellipsoid is WGS84
+// (Maybe all this is because the "Globe" is not available in the tile provider...)
```
I removed that part of the comment. 

But the path that some of the flags, like `showGroundAtmosphere` and like ~20 (!) others take, from https://github.com/CesiumGS/cesium/blob/6cc004aaff586bb59f07f199216ae511014cf5a9/packages/engine/Source/Scene/Globe.js#L1029 through the `GlobeSurfaceTileProvider`, until they end up at https://github.com/CesiumGS/cesium/blob/6cc004aaff586bb59f07f199216ae511014cf5a9/packages/engine/Source/Scene/GlobeSurfaceShaderSet.js#L178 are beyond obscure. I think that this should _definitely_ be revised.

javagl commented on packages/engine/Source/Scene/Model/ImageryFlags.js:
```diff
@@ -0,0 +1,29 @@
+ * <code>number</code>. Users should check for these flags having
+ * a 'truthy' or 'falsy' value.
+ */
+class ImageryFlags {
```
Maybe I'm missing some magic VS Code plugin or so, but ... for a `class`, it's possible to CTRL+Click on the class name e.g. in `@param {ImageryFlags} example` and browse there. For typedef, this is not possible. I can change this to a `typedef`, but am curious whether there is an avantage that outweighs the disadvantage of not being browsable... (I think that _all_ "typedefs" should become class definitions, actually...)

javagl replied:
This "browsability" might be related to the imports (and whether something _can_ be imported without upsetting ESLint), and to some extent, whether that type is a `Standalone` or a `Main.Subtype`. I haven't analyzed the details here. It might be related to https://github.com/CesiumGS/cesium/issues/10455 unter the hood.

javagl commented on packages/engine/Source/Scene/Model/ImageryInput.js:
```diff
@@ -0,0 +1,60 @@
+ *
+ * @private
+ */
+class ImageryInput {
```
See comment above

javagl commented on packages/engine/Source/Scene/Model/ImageryCoverage.js:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @private
+ */
+class ImageryCoverage {
```
For some parts of the code, the first goal was to create building blocks. And nothing is more a building block than a static function. For the "final" structure, there are different degrees of freedom (and styles). But here, I agree that given that `ImageryCoverageComputations` **only** contains `static` functions (and essentially only **one** is **not** `private`), it probably makes sense to move them into `ImageryCoverage`.

javagl commented on packages/engine/Source/Scene/Model/ImageryCoverageComputations.js:
```diff
@@ -0,0 +1,385 @@
+   * @param {number} imageryLevel The imagery level
+   * @returns {number} The validated level
+   */
+  static _validateImageryLevel(imageryProvider, imageryLevel) {
```
Done

javagl commented on packages/engine/Source/Scene/Model/ModelImagery.js:
```diff
@@ -0,0 +1,449 @@
+    this._checkForModifiedImageryConfigurations();
+  }
+
+  // XXX_DRAPING_UPSAMPLING Reset after new scene graph was built
```
That was a leftover from the upsampling experiments. Removed for now.

javagl commented on packages/engine/Source/Scene/Model/ModelImagery.js:
```diff
@@ -0,0 +1,449 @@
+    Check.defined("model", model);
+    //>>includeEnd('debug');
+
+    /**
+     * The model that this instance was created for.
+     *
+     * @type {Model}
+     * @readonly
+     * @private
+     */
+    this._model = model;
```
I'm a huge fan of _clean_ directional dependencies, _clean_ ownership concepts, _clean_ interfaces, _clean_ (SOLID) code in general (and this causes lots and lots of [weltschmerz](https://www.merriam-webster.com/dictionary/weltschmerz)). We already _do_ have circular dependencies _everywhere_. Many of them are not necessary. Many of them should be cleaned up. 

That said: It sould be possble to pass the `Model` to `update`. This means that the `ModelImagery` will have to store the `imageryLayers` of the `Model`. And it will have to make the assumption that the `model` that is passed to the `update` function is _always the same model_. (That may be self-evident, but it is _necessary_ for things to work properly, and it can _not_ be _verified_ sensibly). 

I'll see what I can do...

javagl replied:
Well, given that the `ModelPrimitiveImagery` also stores the `Model` (which is also not nice), this may require a larger refactoring. If this was a strongly requested change, drop me a note. Otherwise, we both have to weep in weltschmerz, and schedule a larger refactoring and cleanup.

javagl commented on packages/engine/Source/Scene/Model/ModelPrimitiveImagery.js:
```diff
@@ -0,0 +1,994 @@
+    Check.defined("runtimePrimitive", runtimePrimitive);
+    //>>includeEnd('debug');
+
+    /**
+     * The model that this instance was created for.
+     *
+     * @type {Model}
+     * @readonly
+     * @private
+     */
+    this._model = model;
+
+    /**
+     * The node that the primitive is attached to
+     *
+     * @type {ModelRuntimeNode}
+     * @readonly
+     * @private
+     */
+    this._runtimeNode = runtimeNode;
+
+    /**
+     * The primitive that this instance was created for.
+     *
+     * @type {ModelRuntimePrimitive}
+     * @readonly
+     * @private
+     */
+    this._runtimePrimitive = runtimePrimitive;
```
See notes above. If this is an actual change request, I can have another look. For now, it's just one of hundreds of circular dependencies.

javagl commented:
I addressed most of the inlined comments from that review pass. I should probably mention (or rather _have_ mentioned) that I did not yet do a throrough self-review, and would certainly have addressed most of the minor things (like missing `@private`s or `return {Types}`) duing that, but ... now it's done, so that's OK. What's still open from the higher-level feedback:

- The `BoundingVolumeDerivation` class: Pulling these functions out of `Implicit3DTileContent` was part of the upsampling preparations, but it's unclear whether this makes sense (beyond "cleanup and structuring"). This could be "undone", if this is preferred.
- The "flat" objects `ImageryFlags`, `ImageryConfiguration`, `ImageryInput`: While these _could_ be converted into typedefs, I see some benefits of things having actual _structures_ and _names_ on the language level. In fact, I'd advocate for changing the existing typedefs into proper classes. (One could consider this as one tiny, tiny, tiny part of https://github.com/CesiumGS/cesium/issues/8359 ... although my gut feeling is that this sort of "cleanups" will hardly ever be tackled...)
- About the "static classes" (now only `ModelImageryMapping`): Many of the static functions have been created in the process of creating building blocks and carving out parts of the existing 700-lines-functions. Now that these blocks exist, they could be moved anywhere. And the classes in particular could be converted into an "Object", as described in the TypeScript docs, but as mentioned in https://github.com/CesiumGS/cesium/pull/12567#discussion_r2105447910 , I'm curious whether this has actual benefits...
- Not storing the `Model` in `ModelImagery` and `ModelPrimitiveImagery`: While I agree that this would be nice _in general_, nearly everything that is _owned by_ the model also _stores_ the model (similar to how nearly everything that is _owned_ by the tileset also _stores_ the tileset). If this was a specific change request, I can have another look. (It would certainly require some restructuring and passing along the model through some call chains, but may be doable...)
- There open points about the `ModelReader` (including the `TODO` comments):  Parts of this class (the undocumented ones) are part of the upsampling preparations/experiments. Specifically, some of the functions are actually not used in the current state. I can remove the functions that are currently unused, but will first have to sync the _current_ state of this PR with my local state that contains the upsampling experiments.

I had a short look whether we have something built-in that could replace the "Bing Labels Only", but there doesn't seem to be anything in the asset depot (or that is easily accessible otherwise).

javagl commented on packages/engine/Source/Scene/Model/ModelReader.js:
```diff
@@ -0,0 +1,762 @@
+
+  /**
+   * @param {ModelComponents.Indices} primitiveIndices
+   * @returns {TypedArray} TODO
```
Added comments there. (Some of these functions are used in the specs of this class, even though the functions are not required for the _current_ state of the draping, but only for upsampling. I can try to separate that, if desired)

javagl commented:
I added the first pass of specs. Until now, they mainly cover the "leaves of call trees", i.e. `static` functions, but also a few ones that rely on actual _state_ and model/tileset loading. There are several dimensions along which the coverage should be more thorough. On the one hand, I'm tempted to create more spec _data_ files (more unit squares!!!) and think that they could be useful elsewhere. On the other hand, there are many paths related to the actual "lifecycle of the draping _itself_" (down to the level of checking reference counters for specific `Imagery` objects). Let's see how far we can get here.

<sub>And regarding the previous build failures: No matter how pedantic you are, there's always a computer that's _more_ pedantic than you 😆 </sub>

javagl commented:
The last commits mainly consisted of adding specs. Unfortunately, one of the specs uncovered an open question (that had been in the code as `TODO_DRAPING` for a while now), namely that of the reference counting for imageries: The `Imagery` has a `referenceCount` property. And this was not maintained properly (skipping some details). The solution is to store the `Imagery` object directly in the `ImageryCoverage` (which I originally tried to avoid, for various reasons - again, skipping some details). It's certainly not ideal to do such a change so late in the process, but given that there now **is** a (very basic) spec for this, and the previous test sandcastles still seem to work, I hope that this turns out to be OK.

ggetz commented:
@javagl I hope you don't mind– I pushed some minor inline documentation updates, mostly to cross link and add more examples.

javagl commented:
@ggetz These are the most important ones. I just opened https://github.com/CesiumGS/cesium/issues/12643 with these (and some details). Further points can be added to this issue, or dedicated new issues, as we see fit.
